### ISSUE TITLE
The option to write an xmp sidecar file "after edit" has unwanted side effects that interfere with file backup and -transfer

### ISSUE THREAD

**Issue by bertwim** [darktable-org/darktable#16676]
### Describe the bug

When writing xmp sidecars "after edit',  DT writes  a new xmp file after an image has been edited in Darkroom. From the code (in image.c)  I understand that DT determines from a hash if an image has been changed. This seems to work ok. 

The issue occurs after DT has been closed and reopened. In that case, DT might also unnecessarily rewrite other xmp sidecar files. I say 'might', because the precise behaviour seems a bit erratic. The issue is with the timestamps. 

When the 'other files'  are rewritten, they have new timestamps, thus appearing to be new files whereas they are only rewrites of their originals. A loss of valuable information that pollutes any (incremental) backup that is made.  
Also, as I have to work from different workstations that are kept in sync with rsync,  there is constant confusion which files are now being really changed and which not. 

The problem is complete when the DT-datebase is cleared first (hence automatically rebuilt with the xmp files): In that case *all* xmp are rewritten, with all having the same timestamp! DT should *never* change the meta data in this case.

The desired behaviour should be, in my opinion:
- Only write xmp if there has been an edit (as the option "after edit" indicates)
- Nothing else. No side effects. 
- No writing of xmp files that have just been read

Regards
Bertwim 

### Steps to reproduce

1. Go to any directory that has images and open DT:   cd ......; darktable . &
2. do a 'ls -l *xmp"   to list all xmp-files. Check the timestamps.
3. Edit one or more images and list the xmp files again
4. Close DT. Then reopen DT again (darktable . &)
5. Carefully compare the timestamps fo the xmp files.
6. Close DT
7. Remove the database (e.g. ~/.config/darktable/library*db) and cache (e.g. ~/.cache/darktable)
8. Reopen DT and again compare timestamps of the xmp files.

### Expected behavior

- write XMP of an image only after an edit  (this is ok).  *Nothing else. No side effects.

### Where did you obtain darktable from?

GitHub nightly

### darktable version

darktable 4.7.0+1000~gd05ae79e9b

### What OS are you using?

Linux

### What is the version of your OS?

OpenSUSE 15.5 

### Are you using OpenCL GPU in darktable?

Yes

### If yes, what is the GPU card and driver?

Irrelvent for this problem. 

### Please provide additional context if applicable. You can attach files too, but might need to rename to .txt or .zip

Problem has been in DT for many years, I think. 
Personally, I have made a local fix that completely avoids the issue.

**Comment by ralfbrown** [comment]
dt also updates sidecars when image metadata changes (e.g. tags, stars, color labels, title, etc.) so that you don't lose that information if the database is lost.

In the particular case of importing into a fresh library, the "import time" metadata changes, which is triggering the rewrite.

**Comment by victoryforce** [comment]
> In the particular case of importing into a fresh library, the "import time" metadata changes, which is triggering the rewrite.

@bertwim Is this explanation satisfactory for you? Can we close this issue?

**Comment by bertwim** [comment]
Please do not close this issue!  It is a bug. Explaining why the behaviour is what it is does not mean there is no mistake! Changing the timestamp when nothing has changed is a bug. It gives misleading information and it interferes with backup-schemes.

**Comment by bertwim** [comment]
I think the basic flaw is that DT has a certain conception of what it means that a picture has been edited. It compares hash values. When the hash has changed, it interprets this as the picture has changed and it writes xmp files accordingly. This conception -evolved as it may seem-, however, is not always correct. When sufficient information is lacking DT is taking an action that it just should not do, i.e. writing new files with new timestamps.  As explained several times, these new timestamps spoil a proper backup. 

In my local copy of DT, I have resolved this issue by a very simple fix, of just a few lines: in an additional option, an xmp file is written *only* when it is being asked for. In my case, I found that writing an xmp  automatically when I export an jpg-file (and only then: no automagically writing of xmp)- is exactly what I needed.
I can immediately see which files have been edited. Timestamps are always meaningful. No unforeseen or unsuspected writing of XMP.  I have tried to get this option into DT but it was denied. I never under stood why.  It was alleged that this is not a correct way of doing but, I do not agree. Maybe this does not fit other peoples workflow, but then just don't choose that option. For me it has worked for many years.   

Regards,
Bertwim

**Comment by wpferguson** [comment]
The XMP file in darktable is meant to be a backup file for the database, so the information in it is what's required to rebuild the database.  In that case the definition of edit is a change to the database, not necessarily a change to the image.

In the rebuild the database from XMP files scenario, no database exists so darktable can't tell if the data  in the XMP files has changed, so it has to rewrite the XMP files to make them match the  database.

**Comment by bertwim** [comment]
Well, everybody is busy explaining how DT works. But that is not the point. I understand how it works. But that doesn't mean that DT is always doing the desired thing!  I explained above that the current practice is interfering  with normal file management procedures, notably DT throws away useful metadata -meaningful timestamps- and hence disturbs a normal incremental backup process which to me is, in an understatement, undesired.

**Comment by logkirk** [comment]
I'm having this issue as well. I agree that this option does not work as expected.

**Comment by wpferguson** [comment]
> this option does not work as expected

That's probably the best description of the "after edit" option for everyone.  

This single "feature" has probably generated more issues than any other and almost all of the issues are about the definition of "edit".

Some of the various definitions of edit are:

- open an image in darkroom and apply the default modules
- open an image in darkroom and apply a module other than the default ones
- adding a tag
- changing the rating
- changing the color label
- changing the metadata (rights, creator, etc)
- grouping
- applying a style (both ways, counts as an edit and doesn't count as an edit)
- change timestamp changing
- and a lot more that I didn't 
- any timestamp changing
- any change to an image

And then we have the issues where users counted on the after edit XMP to be a database backup and it didn't work the way they expected.

IMO this option needs to go away because there isn't a way to satisfy all the users.  

The only way I can see to satisfy this issue, without causing a lot of other issues, is to implement a huge amount of preferences to try and define "edit" along with a huge decision tree to implement the preferences.  Once that's implemented then we get the second wave of preferences that cover all the things we didn't think of the first time.

---

I just thought of another possible way to do this:

The problem with "after edit" is that everyone has a different opinion of what that means so in essence "after edit" is a custom workflow.  The custom workflow tool in darktable is Lua and scripting.  We could:

- add a function to the image datatype to write a sidecar file.
- add an event or events to determine when various aspects of an image change.
- write specific scripts for each use case
- the user sets write sidecar to never and then runs the appropriate script that meets their use case
- the script writes the sidecar as specified in the script.

Advantages:

- No preference changes (other than hopefully remove after edit :smile:)
- Minimal code changes (add a function call and a couple of Lua events)
- Specific scripts tailored to the use case so that we can hopefully satisfy everyone
- Easy to implement new workflows on the fly without waiting until a release

Drawbacks:

- script prejudice (it's a "script", not "real code")

**Comment by logkirk** [comment]
@wpferguson Thanks for your thoughts on this. I can certainly see how this would be tricky to implement owing to different definitions of 'edit'.

> Some of the various definitions of edit are:
> 
> * open an image in darkroom and apply the default modules
> * open an image in darkroom and apply a module other than the default ones
> * adding a tag
> * changing the rating
> * changing the color label
> * changing the metadata (rights, creator, etc)
> * grouping
> * applying a style (both ways, counts as an edit and doesn't count as an edit)
> * change timestamp changing
> * and a lot more that I didn't
> * any timestamp changing
> * any change to an image

For what it's worth, I don't think(?) I have a problem with any of these things causing a sidecar file to be created, except for the "Import time" metadata changing. This seems obviously incorrect, since this means that sidecar files will always be created when importing new images regardless of the setting.

> I just thought of another possible way to do this:

Sounds like an interesting idea, though unfortunately I'm not a seasoned enough Darktable user to have an opinion on the Lua scripting capabilities yet :)

**Comment by bertwim** [comment]
Yeah. It is disappointing that this is not even recognized as an issue that needs fixing.
It is for this reason that I am patching each 'standard' darktable with a fix (only a few lines of code).
 
Regards,
Bertwim

**Comment by wpferguson** [comment]
> Yeah. It is disappointing that this is not even recognized as an issue that needs fixing.

Nobody said it didn't need fixing.  The problem is that we don't know how to fix it for everyone.  It's easy to fix for 1 person and 1 workflow (as you have already discovered).  But, fixing it so that it works for you breaks it for others.

> This seems obviously incorrect, since this means that sidecar files will always be created when importing new images regardless of the setting.

If you "after edit" set and you import a new set of images, then no sidecars are created (I just tested and it seems that's not entirely correct.  Some DNG and TIF files did generate XMPs, but I also observed error messages in the console about forward matrix to it may have something to do with that).

As far as timestamps, I can make a use case for updating the import timestamp.  If I have 1000 images and I update 5 random XMPs, then I can sort by import time to find the updated images, versus searching through 1000 images trying to find the updated ones.

So, now we have another set of variables to consider, i.e. how should timestamps be handled?  IMO if I'm syncing edits between multiple machines then I think the XMP files should be the same.  But, I can make use cases for them not being the same.  And, I'm not sure darktable can handle keeping the XMP files the same.
